### ISSUE TITLE
false negative if_same_then_else / new lint catch wrongly ordered struct init shorthand

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#6352]
Upstream rustc issue: https://github.com/rust-lang/rust/issues/79210

This is definitely also clippy territory!
We have the [if_same_than_else](https://rust-lang.github.io/rust-clippy/master/index.html#if_same_then_else) lint which probably should catch the upstream example, but I wonder if we could have a lint that catches the field ordering being suspicious.

I tried this code:

```rust
#[derive(Debug)]
struct S {
    a: i32,
    b: i32,
}

fn main() {
    let a = 0_i32;
    let b = 1_i32;
    let s: S = if false {
        S { a, b }
    } else {
        S { b, a } /* wanted: S {a: b, b: a}*/
    };
    println!("{:?}", s);
}
```

The problem is that `S { a, b }` and `S { b, a }` is essentially the same, but the `if_same_then_else` lint did not see this (probably because of the field ordering).
Alternatively, I wonder if it makes sense to have a lint that generally warns we use the shorthand init pattern 
````rust
struct S { a: i32, b: i32, c: i32 }
let a = 1;
let b = 2;
let c = 3;
let s = S { c, a, b }
````
where the struct members are not listed in the order that they are defined in the struct definition.

`clippy 0.0.212 (fe98231 2020-11-19)`

**Comment by GG-Sheridan** [comment]
Should this issue be labeled as Good First Issue?

We are doing research about how to help newcomers onboarding in an open source project and how to accelerate the issue resolving process.
We have developed a tool to automatically label the Good First Issue (GFI), i.e., the issue which is good for newcomers. 
We first extract 72 features from different dimensions of issue reports, such as readability, semantic meaning, associated code snippets, etc, and experimentally investigate to what extent these features correlate with GFI. Results turn out that a large portion of these features do have a medium to large correlation with GFI. 
We then build a machine learning model based on these features to predict GFI. The experiments on a dataset of 12k issues from Github show that 0.90 AUC can be achieved. 

Our tool shows that this issue has a very large possibility to be a GFI. 
Could you please tell us whether it should be labeled as a GFI?
And if you have any comments or suggestions with this tool, could you please tell us? 
Thank you so much.

**Comment by Y-Nak** [comment]
As @matthiaskrgr suggested, I'll add the new lint to fix this issue (and the lint is quite useful regardless of this issue itself).
@rustbot claim
